src/lib/storage/table.cpp

```diff
@@ -17,53 +19,103 @@
+    }
+  }
+
+  _chunks.back()->append(values);
 }
 
+bool Table::is_new_chunk_needed() { return _chunks.back()->size() >= _max_chunk_size; }
+
+uint16_t Table::column_count() const { return _column_names.size(); }
+
 uint64_t Table::row_count() const {
-  // Implementation goes here
-  return 0;
+  // Since we cannot delete data, all old chunks will have maximum chunk size
+  return ((_chunks.size() - 1) * _max_chunk_size) + _chunks.back()->size();
 }
 
-ChunkID Table::chunk_count() const {
-  // Implementation goes here
-  return ChunkID{0};
-}
+ChunkID Table::chunk_count() const { return (ChunkID)_chunks.size(); }
 
 ColumnID Table::column_id_by_name(const std::string& column_name) const {
-  // Implementation goes here
-  return ColumnID{0};
+  auto search = _column_name_mapping.find(column_name);
+  DebugAssert(search != _column_name_mapping.end(), "Could not find column_name");
+  return search->second;
 }
 
-uint32_t Table::max_chunk_size() const {
-  // Implementation goes here
-  return 0;
-}
+uint32_t Table::max_chunk_size() const { return _max_chunk_size; }
 
-const std::vector<std::string>& Table::column_names() const {
-  throw std::runtime_error("Implement Table::column_names()");
-}
+bool Table::is_empty() { return _chunks.size() == 1 && _chunks.front()->size() == 0; }
 
-const std::string& Table::column_name(ColumnID column_id) const {
-  throw std::runtime_error("Implement Table::column_name");
-}
+const std::vector<std::string>& Table::column_names() const { return _column_names; }
 
-const std::string& Table::column_type(ColumnID column_id) const {
-  throw std::runtime_error("Implement Table::column_type");
-}
+const std::string& Table::column_name(ColumnID column_id) const { return _column_names.at(column_id); }
+
+const std::string& Table::column_type(ColumnID column_id) const { return _column_types.at(column_id); }
+
+Chunk& Table::get_chunk(ChunkID chunk_id) { return *_chunks.at(chunk_id); }
+
+const Chunk& Table::get_chunk(ChunkID chunk_id) const { return *_chunks.at(chunk_id); }
+
+void Table::compress_chunk(ChunkID chunk_id) {
+  DebugAssert(chunk_id < _chunks.size(), "ChunkID out of range");
+  const auto chunk = _chunks[chunk_id];
+  auto compressed_chunk = std::make_shared<Chunk>();
+  // create new empty chunk
+  _chunks.push_back(compressed_chunk);
 
-Chunk& Table::get_chunk(ChunkID chunk_id) { throw std::runtime_error("Implement Table::get_chunk"); }
+  // number of threads needed
+  auto n_columns = chunk->column_count();
+  std::vector<std::thread> threads(n_columns);
+  std::vector<std::shared_ptr<BaseSegment>> dictionary_segments(n_columns);
+  // to enable threads to access specific position in vector
+  dictionary_segments.resize(n_columns);
 
-const Chunk& Table::get_chunk(ChunkID chunk_id) const { throw std::runtime_error("Implement Table::get_chunk"); }
+  for (size_t segment = 0; segment < dictionary_segments.size(); ++segment) {
+    threads.push_back(std::thread([&] (size_t seg) {
```

Comment:
catching the whole scope in the lambda is error prone. actually only dictionary_segment is needed here

Reply:
Yeah, we will probably steal from your implementation here and rather just hand over the relevant elements explicitly

Reply:
Fixed in next commit. Is much nicer now. Thanks